qkit/drivers/ZHInst_Virtual_VNAv2.py

```diff
@@ -0,0 +1,163 @@
+        self.uhfqa.nodetree.sigins[1].range(expected_voltage)
+
+    @QkitFunction
+    def get_freqpoints(self) -> np.ndarray:
+        return self._frequencies
+
+    @QkitFunction
+    def get_tracedata(self, RealImag = None) -> tuple((np.ndarray, np.ndarray)):
+        if not RealImag:
+            signal = (self._i_q_data[:, 0] + 1.0j * self._i_q_data[:, 1])
+            return np.abs(signal), np.angle(signal)
+        else:
+            return self._i_q_data[:, 0], self._i_q_data[:, 1]
+
+    @QkitFunction
+    def _get_shfsg_frequencies(self) -> tuple((float, float)):
+        """
+        Returns (SHF-Upconversion center frequency, Digital Modulation Frequency)
+        """
+        shfsg_out_freq = self.get_freqpoints()[0] - MINIMUM_UHFQA_FREQUENCY
+        base_mod = int(shfsg_out_freq / 0.5e9) * 0.5e9
+        return base_mod, shfsg_out_freq - base_mod
+
+    @QkitFunction
+    def pre_measurement(self):
+        self.uhfqa.nodetree.qa.result.source(2)
+        self.uhfqa.nodetree.sigouts[0].on(1)
+        self.uhfqa.nodetree.sigouts[1].on(1)
+
+        self.uhfqa.nodetree.awg.outputs[1].amplitude(-1.0)
+        self.uhfqa.arm(length=self.repetitions, averages=1)
+        # Configure output
+        base, digital = self._get_shfsg_frequencies()
+        out_channel = self.shfsg.nodetree.sgchannels[7]
+        synth = self.shfsg.nodetree.synthesizers[out_channel.synthesizer()]
+
+        out_channel.output.range(self.output_power)
+        synth.centerfreq(base)
+        out_channel.awg.modulation.enable(1)
+        out_channel.oscs[0].freq(digital)
+
+        self.uhfqa.compile_program(
+            sequence_type=SequenceType.PULSED_SPEC,
+            period=self.period,
+            repetitions=self.repetitions,
+            trigger_mode=TriggerMode.NONE,
+            alignment=Alignment.START_WITH_TRIGGER
+        )
+        
+        self.shfsg.nodetree.sgchannels[7].output.on(1)
+
+        self.uhfqa.nodetree.qa.result.source(2)
+        self.uhfqa.nodetree.sigouts[0].on(1)# Enable the outputs
+        self.uhfqa.nodetree.sigouts[1].on(1)
+        self.uhfqa.nodetree.awg.outputs[1].amplitude(-1.0)
+
+        self._i_q_data = np.zeros(shape=(len(self.get_freqpoints()), 2), dtype=complex)
+
+    @QkitFunction
+    def start_measurement(self):
+        print("Run")
+        # Configure upper side band modulation
+        out_channel = self.shfsg.nodetree.sgchannels[7]
+        out_channel.sine.oscselect(0)
+        out_channel.sine.i.enable(True)
+        out_channel.sine.i.sin.amplitude(0.0)
+        out_channel.sine.i.cos.amplitude(0.9 * self.amplitude_factor)
+        out_channel.sine.q.enable(True)
+        out_channel.sine.q.sin.amplitude(0.9 * self.amplitude_factor)
+        out_channel.sine.q.cos.amplitude(0.0)
+        self.uhfqa.arm(length=self.repetitions, averages=1)
+
+        uhfqa_frequencies = self.get_freqpoints() - (self.get_freqpoints()[0] - MINIMUM_UHFQA_FREQUENCY)
```

Comment:
Bug: If SHFSG frequency is not MINIMUM_UHFQA_FREQUENCY below lowest frequency point this calculation is incorrect.
Use Actual SHFSG frequency as reference instead.
Bonus: Check if UHFQA frequency range is exceeded.